pkg/spdk/disk/types.go

```diff
@@ -0,0 +1,139 @@
+)
+
+type BlockDiskSubsystem string
+
+const (
+	BlockDiskSubsystemVirtio = BlockDiskSubsystem("virtio")
+	BlockDiskSubsystemPci    = BlockDiskSubsystem("pci")
+	BlockDiskSubsystemNvme   = BlockDiskSubsystem("nvme")
+	BlockDiskSubsystemScsi   = BlockDiskSubsystem("scsi")
+)
+
+const (
+	UioDriverUioPciGeneric = "uio_pci_generic"
+)
+
+type BlockDiskType string
+
+const (
+	BlockDiskTypeDisk = BlockDiskType("disk")
+	BlockDiskTypeLoop = BlockDiskType("loop")
+)
+
+type DiskStatus struct {
+	Bdf          string
+	Type         string
+	Driver       string
+	Vendor       string
+	Numa         string
+	Device       string
+	BlockDevices string
+}
+
+func GetDiskDriver(diskDriver, diskPathOrBdf string) (BlockDiskDriver, error) {
+	if isBDF(diskPathOrBdf) {
+		return getDiskDriverForBDF(diskDriver, diskPathOrBdf)
+	}
+
+	return getDiskDriverForPath(diskDriver, diskPathOrBdf)
+}
+
+func getDiskDriverForBDF(diskDriver, bdf string) (BlockDiskDriver, error) {
+	executor, err := helperutil.NewExecutor(commonTypes.ProcDirectory)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get the executor for disk driver detection")
+	}
+
+	output, err := spdksetup.GetDiskStatus(bdf, executor)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get disk status for BDF %s", bdf)
+	}
+
+	var diskStatus DiskStatus
+	err = json.Unmarshal([]byte(output), &diskStatus)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to unmarshal disk status for BDF %s", bdf)
+	}
+
+	switch BlockDiskDriver(diskDriver) {
+	case BlockDiskDriverAuto:
+		diskPath := ""
+		if BlockDiskDriver(diskDriver) == BlockDiskDriverAuto && BlockDiskDriver(diskStatus.Driver) != UioDriverUioPciGeneric {
+			devName, err := util.GetDevNameFromBDF(bdf)
+			if err != nil {
+				return "", errors.Wrapf(err, "failed to get device name from BDF %s", bdf)
+			}
+			diskPath = fmt.Sprintf("/dev/%s", devName)
+		}
+		return getDriverForAuto(diskStatus, diskPath)
+	case BlockDiskDriverAio, BlockDiskDriverNvme, BlockDiskDriverVirtioScsi, BlockDiskDriverVirtioBlk:
+		return BlockDiskDriver(diskDriver), nil
+	default:
+		return "", fmt.Errorf("unsupported disk driver %s for BDF %s", diskDriver, bdf)
+	}
+}
+
+func getDriverForAuto(diskStatus DiskStatus, diskPath string) (BlockDiskDriver, error) {
+	switch BlockDiskDriver(diskStatus.Driver) {
+	case BlockDiskDriverNvme:
+		return BlockDiskDriver(diskStatus.Driver), nil
+	case BlockDiskDriverVirtioPci:
+		subsystems, err := util.GetBlockDiskSubsystems(diskPath)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to get disk subsystems for %s", diskPath)
+		}
+
+		if slices.Contains(subsystems, string(BlockDiskSubsystemVirtio)) && slices.Contains(subsystems, string(BlockDiskSubsystemPci)) {
+			diskDriver := BlockDiskDriver(BlockDiskDriverVirtioBlk)
+			if slices.Contains(subsystems, string(BlockDiskSubsystemScsi)) {
+				diskDriver = BlockDiskDriver(BlockDiskDriverVirtioScsi)
+			}
+			return diskDriver, nil
```

Comment:
Would you explain the relationship between these drivers?

Reply:
SPDK supports various types of disks, including NVMe, virtio-blk, and virtio-scsi.

NVMe disks can be managed by either NVMe bdev or AIO bdev.
VirtIO disks can be managed by virtio-blk, virtio-scsi, or AIO bdev.
To use the correct bdev, you need to identify the disk type.

Here's how to identify the disk type:

- If a block device uses the subsystems virtio and pci, it's a virtio-blk disk.
- If it uses the subsystems virtio, pci, and scsi, it's a virtio-scsi disk.